store/src/main/java/org/apache/rocketmq/store/CommitLog.java

```diff
@@ -623,27 +646,22 @@ public long getBeginTimeInLock() {
                     mappedFile = this.mappedFileQueue.getLastMappedFile(0);
                     if (null == mappedFile) {
                         // XXX: warn and notify me
-                        log.error("create mapped file2 error, topic: " + msg.getTopic() + " clientAddr: " + msg.getBornHostString());
-                        beginTimeInLock = 0;
-                        return CompletableFuture.completedFuture(new PutMessageResult(PutMessageStatus.CREATE_MAPEDFILE_FAILED, result));
+                        log.error("create mapped file error, topic: " + msg.getTopic() + " clientAddr: " + msg.getBornHostString());
+                        throw new Exceptions.CreateMappedfileFailedException(result);
                     }
                     result = mappedFile.appendMessage(msg, this.appendMessageCallback);
                     break;
                 case MESSAGE_SIZE_EXCEEDED:
                 case PROPERTIES_SIZE_EXCEEDED:
-                    beginTimeInLock = 0;
-                    return CompletableFuture.completedFuture(new PutMessageResult(PutMessageStatus.MESSAGE_ILLEGAL, result));
+                    throw new Exceptions.MessageIllegalException(result);
```

Comment:
Is it possible that throw an Exception will affect performance?

Reply:
It requires to create an extra object (the exception itself) and bring the stack trace of current thread, I've not any measure about performance loss. Do you think this is critical?

Reply:
from https://www.baeldung.com/java-exceptions-performance throwing exception is more costly than returning an error, but how many times do you expect the exception is raised? 
The condition under which is raised are:
- message size too big
- the properties are too big
- mapped file doesn't exists.

Do you think they happen so often to generate a performance problem in throwing exception instead of returning error code?

Reply:
Hi, I moved from throwing exceptions to returning a couple with error or result